src/toil/batchSystems/kubernetes.py

```diff
@@ -0,0 +1,668 @@
+# Copyright (C) 2019 Regents of the University of California
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Batch system for running Toil workflows on Kubernetes.
+
+Ony useful with network-based job stores, like AWSJobStore.
+
+Within non-priveleged Kubernetes containers, additional Docker containers
+cannot yet be launched. That functionality will need to wait for user-mode
+Docker
+"""
+
+from __future__ import absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import str
+
+import base64
+import datetime
+import getpass
+import kubernetes
+import logging
+import os
+import pickle
+import pytz
+import subprocess
+import sys
+import uuid
+import time
+
+from kubernetes.client.rest import ApiException
+from six.moves.queue import Empty, Queue
+
+from toil import applianceSelf, customDockerInitCmd
+from toil.batchSystems.abstractBatchSystem import (AbstractBatchSystem,
+                                                   BatchSystemLocalSupport)
+from toil.resource import Resource
+
+logger = logging.getLogger(__name__)
+
+
+class KubernetesBatchSystem(BatchSystemLocalSupport):
+
+    @classmethod
+    def supportsAutoDeployment(cls):
+        return True
+
+    @classmethod
+    def supportsWorkerCleanup(cls):
+        return False
+   
+    def __init__(self, config, maxCores, maxMemory, maxDisk):
+        super(KubernetesBatchSystem, self).__init__(config, maxCores, maxMemory, maxDisk)
+
+        try:
+            # Load ~/.kube/config
+            kubernetes.config.load_kube_config()
+        except TypeError:
+            raise RuntimeError('Could not load Kubernetes configuration. Does ~/.kube/config or $KUBECONFIG exist?')
+
+        # Find all contexts and the active context.
+        # The active context gets us our namespace.
+        contexts, active_context = kubernetes.config.list_kube_config_contexts()
+        if not contexts:
+            raise RuntimeError("No Kubernetes contexts available in ~/.kube/config or $KUBECONFIG")
+            
+        # Identify the namespace to work in
+        self.namespace = active_context.get('context', {}).get('namespace', 'default')
+
+        # Create a prefix for jobs, starting with our username
+        self.jobPrefix = '{}-toil-{}-'.format(getpass.getuser(), uuid.uuid4())
```

Comment:
Nit: Might want to limit the length of this?

Reply:
I want to leave it as the full UUID, I think. [Kubernetes says we can have 253 characters](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) and a UUID is pretty short compared to that, and guaranteed to be unique. If we limit it to something more like a Git commit hash, we start getting non-astronomically-improbably collision probabilities (although you'd still have to run a billion Toil workflows at once on the same cluster).

We could cut the user name to some non-excessively-long substring. The other problem is keeping non-ASCII username characters (supported on some not-quite-POSIX systems) out of Kubernetes. I think I might make another PR for this.